// enthusiast/index.js
import cookie from 'cookie';
import parser from 'accept-language-parser';

const defaultSettings = {
	cookieName: '__enthusiast_locale',
	defaultLocale: 'en',
	excludedRoutes: [],
	localeParameterIndex: 1,
	locales: ['en']
};

export default function enthusiast(settings = defaultSettings) {
	settings = {...defaultSettings, ...settings};

	return (req, res, next) => {
		const path = req.path;

		for (const route of settings.excludedRoutes) {
			if (path.startsWith(route)) {
				return next();
			}
		}

		let requestLocale = settings.defaultLocale;

		const cookies = cookie.parse(req.headers.cookie || '');
		const params = path.split('/');

		if (path.length <= 1) {
			requestLocale = parser.pick(settings.locales, req.headers['accept-language'], { loose: true });

			if (!!cookies[settings.cookieName]) {
				for (const locale of settings.locales) {
					if (locale === cookies[settings.cookieName]) {
						requestLocale = locale;
						break;
					}
				}
			}

			req.locale = requestLocale;

			if (!cookies[settings.cookieName] || settings.cookieName !== requestLocale) {
				const localeCookie = cookie.serialize(settings.cookieName, requestLocale);
				res.setHeader('Set-Cookie', localeCookie);
			}

			return next();
		}

		const pathLocale = params[settings.localeParameterIndex];

		if (!requestLocale.match('^[a-z]{2}(-[A-Z]{2})*$')) {
			requestLocale = settings.defaultLocale;
		}

		requestLocale = pathLocale;
		req.locale = requestLocale;

		if (!cookies[settings.cookieName] || settings.cookieName !== requestLocale) {
			const localeCookie = cookie.serialize(settings.cookieName, requestLocale);
			res.setHeader('Set-Cookie', localeCookie);
		}

		return next();
	};
};
